relay/circuit-v2.md

```diff
@@ -0,0 +1,193 @@
+termination in the same protocol has made it very hard to on provide
+relay service on demand, decoupled with whether _client_ functionality
+is supported by the host.
+
+In order to address this problem, we have splt the protocol into the
+`hop` and `stop` subprotocols. This allows us to always enable the
+client-side functionality in a host, while providing the option to
+later mount the relay service in public hosts, _after_ the
+reachability of the host has been determined through AutoNAT.
+
+## The Protocol
+
+### Interaction
+
+The following diagram illustrates the interaction between three peers,
+_A_, _B_, and _R_, in the course of establishing a relayed connection.
+Peer _A_ is a private peer, which is not publicly reachable; it
+utilizes the services of peer _R_ as the relay.  Peer _B_ is another
+peer who wishes to connect to peer _A_ through _R_.
+
+![Circuit v2 Protocol Interaction](circuit-v2.png)
+
+The first part of the interaction is _A_'s reservation of a relay slot
+in _R_.  This is accomplished by opening a connection to _R_ and
+sending a `RESERVE` message in the `hop` protocol; if the reservation
+is successful, the relay responds with a `STATUS:OK` message and
+provides _A_ with a reservation voucher.
+
+The second part of the interaction is the establishment of a circuit
+switch connection from _B_ to _A_ through _R_.  It is assumed that _B_
+has obtained a circuit multiaddr for _A_ of the form
+`/p2p/QmR/p2p-circuit/p2p/QmA` out of band using some peer discovery
+service (eg. the DHT or a rendezvous point).
+
+In order to connect to _A_, _B_ then conncts to _R_, opens a `hop`
+protocol stream and sends a `CONNECT` message to the relay.  The relay
+verifies that it has a reservation and connection for _A_ and opens a
+`stop` protocol stream to _A_, sending a `CONNECT` message.
+
+Peer _A_ then responds to the relaywith a `STATUS:OK` message, which
+responds to _B_ with a `STATUS:OK` message in the open `hop` stream
+and then proceeds to bridge the two streams into a relayed connection.
+The relayed connection flows in the `hop` stream between the
+connection initiator and the relay and in the `stop` stream between
+the relay and the connection termination point.
+
+### Hop Protocol
+
+TBD
+
+### Stop Protocol
+
+TBD
+
+### Reservation Vouchers
+
+TBD
+
+## Protobuf
+
+```
+message HopMessage {
+  enum Type {
+    RESERVE = 0;
+    CONNECT = 1;
+    STATUS = 2;
+  }
+
+  required Type type = 1;
+
+  optional Peer peer = 2;
+  optional Reservation reservation = 3;
+  optional Limit limit = 4;
+
+  optional Status status = 5;
+}
+
+message StopMessage {
+  enum Type {
+    CONNECT = 0;
+    STATUS = 1;
+  }
+
+  required Type type = 1;
+
+  optional Peer peer = 2;
+  optional Limit limit = 3;
+
+  optional Status status = 4;
+}
+
+message Peer {
+  required bytes id = 1;
+  repeated bytes addrs = 2;
+}
+
+message Reservation {
+  optional int64 expire = 1;  // Unix expiration time (UTC)
```

Comment:
```suggestion
  optional uint64 expire = 1;  // Unix expiration time (UTC)
```

Why not use an unsigned integer?

https://developers.google.com/protocol-buffers/docs/proto#scalar

While we don't need the precision, for the sake of convenience, we could as well consider using a `Timestamp`, no?

https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Timestamp

Reply:
no, we don't use any of the custom google types; that's a very bad idea for portability.

Reply:
The reason for signed integer here is again programming convenience, but we could use unsigned as well.

Reply:
turned into unsigned.